Handle disputable metadata

fabriziovigevani opened the pull request:
closes #369 

Needs https://github.com/aragon/court-subgraph/pull/67 to be deployed first.

Uses [Aragon connect](https://connect.aragon.org/) to describe the dispute metadata for disputes created by agreements (disputables):

- `disputedActionRadspec` (long description) => Built with `describeScript`
- `disputedActionText` (short description) => Built by using the `disputableAppName` and terminal action's role (e.g. `Tokens: Mint tokens`)
- `disputedActionURL` URL of the frontend where the disputed action is taking place.

Old disputes (eg Precedence campaign disputes) are still processed in the same way as before.

We still need to handle Conviction Voting which will be done after this is merged.

⚠️ Caveats

When fetching disputables metadata, it can take up to 5 seconds for the dispute to load (A cache for storing already described disputes was implemented)


**Previews**
####  Taken from PieDAO's Org
![Screen Shot 2020-08-05 at 5 16 42 PM](https://user-images.githubusercontent.com/22663232/89459713-78ee0e00-d73f-11ea-9f0b-046c3333081b.png)



#### Taken from address book used for app mining
![Screen Shot 2020-08-05 at 5 15 18 PM](https://user-images.githubusercontent.com/22663232/89459591-43e1bb80-d73f-11ea-94f4-259b15467b03.png)


#### Where the terminal action is not the immediate step
![Screen Shot 2020-07-08 at 12 52 03 PM](https://user-images.githubusercontent.com/22663232/86954230-ff590500-c12b-11ea-9f45-e60251ea2385.png)
![Screen Shot 2020-08-05 at 5 13 07 PM](https://user-images.githubusercontent.com/22663232/89459403-f9603f00-d73e-11ea-9316-40e7fabbaa7c.png)

#### When the dispute does not have an executable action
![Screen Shot 2020-08-05 at 5 23 34 PM](https://user-images.githubusercontent.com/22663232/89460293-6c1dea00-d740-11ea-92d2-50997e92fa78.png)

sohkai commented:
> When fetching disputables metadata, it can take up to 5 seconds for the dispute to load (A cache for storing already described disputes was implemented)

We should probably come up with a loading state for this! Is this due to the IPFS look up or the radspec look up?

> Where the terminal action is not the immediate step

I think in this case, we would likely still want some explanation that it's going through a more complex process (e.g. your Agreement's rules may be different if a vote is involved than not).

How does it handle the terminal action doing multiple things (E.g. separate mint and burn actions?)

fabriziovigevani commented:
> We should probably come up with a loading state for this!

We do have a loading state for the dispute detail. Are you suggesting we have a different one for this?

![Screen Shot 2020-07-08 at 3 45 00 PM](https://user-images.githubusercontent.com/22663232/86960680-43e99e00-c136-11ea-8ee9-57912c1f5795.png)

>  Is this due to the IPFS look up or the radspec look up?

It's due to the radspec lookup. 


> I think in this case, we would likely still want some explanation that it's going through a more complex process (e.g. your Agreement's rules may be different if a vote is involved than not).

Yeah this makes sense. We can probably do something similar as the new vote | delay descriptions?

> How does it handle the terminal action doing multiple things (E.g. separate mint and burn actions?)

It just prints them separated by a new line. Haven't tried with terminal actions which correspond to different methods but i assume it should handle them the same way.
![Screen Shot 2020-07-08 at 12 49 54 PM](https://user-images.githubusercontent.com/22663232/86961470-a98a5a00-c137-11ea-8d47-a4f9dec1899b.png)

sohkai commented:
> It's due to the radspec lookup.

Interesting, this is surprisingly long even with Connect! Maybe we should have just a loading state for the description, since it's just this part that takes so long.

Let's use the round of testing or precedence campaign disputes to see what this is like on a live dispute and evaluate.

> We can probably do something similar as the new vote | delay descriptions?

Yes, we may have a couple things from the client we could bring over here as we complete them there (multi transaction stepper, transaction path viewer).

sohkai commented on src/disputables/index.js:
```diff
@@ -0,0 +1,261 @@
+ * Builds the URL corresponding to the disputed action. (e.g the URL of a vote)
+ * @param {String} organization Address of the organization
+ * @param {String} appAddress Address of the app
+ * @param {String} entity Entity where the action is being disputed (e.g vote, delay, etc)
```
Maybe `entityPath` is a slightly more descriptive name :)

sohkai commented on src/disputables/index.js:
```diff
@@ -0,0 +1,261 @@
+        .split('-')
+        .map(capitalize)
+        .join(' ')
+    : 'System app' // TODO: Update to more meaningfull name?
```
Oh interesting, we should have the name available for these system apps!

If they're not available, please add an issue to Connect!

In this case, I think we might want to put something like `Unknown app (<app id>)` instead.

fabriziovigevani replied:
Just checked the subgraphs again and i can see that they are available on the Aragon Mainnet subgraph but not on the Rinkeby subgraph. Just notified @0xGabi  about it.

sohkai commented on src/disputables/index.js:
```diff
@@ -0,0 +1,261 @@
+      const result = await disputableAppContract[fn](disputableActionId)
+      const evmScript = result[scriptPosition]
+
+      // Get long and short description corrsponding to the disputed action.
```
```suggestion
      // Get long and short description corresponding to the disputed action.
```

sohkai commented on src/disputables/index.js:
```diff
@@ -0,0 +1,261 @@
+      return disputedActionDescription
+    }
+  } catch (err) {
+    console.error('Error describing disputable action', err)
```
Should we also return with a description that mentions it errorred?

It might also be useful to log this error to sentry with something like [this](https://github.com/aragon/aragon/blob/master/src/sentry.js#L11) so we can see if this happens a lot to people.

fabriziovigevani replied:
> Should we also return with a description that mentions it errorred?

Good question. Currently we are just returning `No description`  as description but maybe makes sense to tell that it actually errored yeah.

sohkai replied:
Probably not worth showing the full error message.

I think something along the lines of "Failed to fetch description" is fine as long as we're receiving alerts (so we can later debug it rather than ask the user specifically for what they saw).

sohkai commented on src/disputables/index.js:
```diff
@@ -0,0 +1,261 @@
+
+  // Get disputed action short text
+  // Find method corresponding to the function of the disputed action.
+  const method = await findAppMethodFromIntent(app, terminalAction)
```
😅 I didn't know this existed on the connect export list (it doesn't appear to be documented).

I can roughly figure out what it's doing, but I'm not sure it's actually meant to be exported as an API from Connect! Did Gabi mention this was available?

fabriziovigevani replied:
haha yes when i was working on this i mentioned @0xGabi what data we needed to have, and he mentioned about this function. Maybe he exposed it so we could use it in this case.

sohkai commented on src/disputables/index.js:
```diff
@@ -0,0 +1,261 @@
+    (await describeScript(evmScript, apps, org.provider)) || []
+
+  // In order to get the terminal action/s we must search through the `transactionRequests` children (if any)
+  const terminalActions = getTerminalActions(transactionRequests)
```
I wonder if we should re-organize the rest of the function a bit:

- Bail early if `terminalActions` is empty
- Push `disputedActionText` and `terminalAction` together, since the comment about the first action only applies to `disputedActionText`

fabriziovigevani replied:
Much better

fabriziovigevani commented on src/components/Disputes/DisputeInfo.js:
```diff
@@ -121,7 +121,13 @@ const DisputeInfo = React.memo(function({
+                    label="Description"
+                    value={disputedActionRadspec}
+                    css={`
+                      overflow-wrap: anywhere;
```
`overflow-wrap: break-word` strangely doesn't break long words. 

This is a not so common scenario when the long word corresponds to an `appId`
![Screen Shot 2020-07-23 at 12 06 58 AM](https://user-images.githubusercontent.com/22663232/88248918-26115280-cc79-11ea-8e41-a7ac5ffa234b.png)

We could also add the `word-break: break-word` style since `overflow-wrap: anywhere` is [not supported by some browsers](https://developer.mozilla.org/en-US/docs/Web/CSS/overflow-wrap#Browser_compatibility)

sohkai replied:
Sounds good, let's add it as well and adjust based on how it looks with real usage!

fabriziovigevani commented on src/lib/web3-utils.js:
```diff
@@ -139,7 +139,7 @@ export function isLocalOrUnknownNetwork(chainId = env('CHAIN_ID')) {
 }
 
 export function hexToAscii(hexx) {
```
Ah yeah probably better to use `ehters` utils

fabriziovigevani commented on src/web3-contracts.js:
```diff
@@ -14,26 +14,23 @@ export function useContract(address, abi, signer = true) {
+export function getContract(
+  address,
+  abi,
+  provider = new Providers.JsonRpcProvider(defaultEthNode)
```
Yes, that's a good idea.

fabriziovigevani commented on src/utils/dispute-utils.js:
```diff
@@ -43,9 +44,38 @@ export const transformRoundDataAttributes = round => {
+    return [dispute.disputable.title]
+  }
+
+  const decodedMetadata = hexToAscii(dispute.metadata)
```
Ah, we changed the type of the `metadata` attribute for the `dispute` entity in the subgraph (from `string` to `bytes`)  
See https://github.com/aragon/court-subgraph/pull/67/files#diff-e37ced9649dd2723e16a424f49db9193R74 

Though is only deploy to the staging subgraph. You can see all the descriptions messed up here https://court-staging.aragon.org/disputes 😄

fabriziovigevani commented:
@sohkai Added the missing bits:

- Loading state for disputed action description exclusively
   **Notes on this:** 
   Moved the rendering of the dispute data into a separate component `DisputeInfoContent`, you can see the commit [HERE](https://github.com/aragon/court-dashboard/pull/373/commits/522724b01fbb1e359946a0590d039da4967866bb) 
  
   Decided also to move the describing of the disputed action (previously on `useDispute` hook) to the component in charge to render it `DisputeInfoContent` using a `useDisputeAction` hook . The reason for this is that we already have a `fetching` indicator for when the dispute is processed in the `useDispute` hook, and adding another loading variable just for this didn't feel right. Relevant commit [HERE](https://github.com/aragon/court-dashboard/pull/373/commits/3c5e2062256ce4e8cd8431459c782c1f233d2bd8)

  Also, added some more  not so relevant files changed for deduping our Loading components that i should have done on a separate PR (sorry for that), here's the commit https://github.com/aragon/court-dashboard/pull/373/commits/e1a8f16cd3a826ee55d5d828e258cf8fd6b9f239

- Return `Failed to fetch description` when there's an error in the process, and log with Sentry.

- Describe all the steps when it's a multi step execution path

**Preview**

Loading
![Screen Shot 2020-07-29 at 6 06 19 PM](https://user-images.githubusercontent.com/22663232/88854769-2a34f700-d1c8-11ea-8338-c6411da7c9d4.png)

Multi step

![Screen Shot 2020-07-29 at 6 06 25 PM](https://user-images.githubusercontent.com/22663232/88854784-33be5f00-d1c8-11ea-9c48-831e6a271b35.png)

sohkai commented on src/disputables/index.js:
```diff
@@ -226,17 +230,16 @@ function buildDisputedActionText(app, method) {
 
 /**
  * Builds the disputed action radspec description
+ * Assumes array of transactions not empty
```
Just curious about the motivation for this change. Can we safely assume this, and are we looking to throw an error (instead of defaulting to an empty string) if it's not the case?

fabriziovigevani replied:
>  Can we safely assume this

Yes we can assume this since we are now early returning when the transactions array is empty 
See [here](https://github.com/aragon/court-dashboard/pull/373/files/e1a8f16cd3a826ee55d5d828e258cf8fd6b9f239..f9d01ba38e89367da5f80b9cac81c80a74693849#diff-07f7c052c786459c4930bc8491c43713R152)

> and are we looking to throw an error

Ah yeah forgot to return an error description here as well instead of the  `No description`.

sohkai commented on src/disputables/index.js:
```diff
@@ -190,6 +192,27 @@ function getTerminalActions(transactions) {
+ * @param {Array} apps Array of all the org's apps
+ * @returns {Array} Execution path
+ */
+function contexctualizeExecutionPath(transactions, apps) {
```
```suggestion
function contextualizeExecutionPath(transactions, apps) {
```

😄

sohkai commented on src/disputables/index.js:
```diff
@@ -190,6 +192,27 @@ function getTerminalActions(transactions) {
+function contexctualizeExecutionPath(transactions, apps) {
+  return transactions.map(transaction => {
+    const app = apps.find(app => addressesEqual(app.address, transaction.to))
+    const appName = app ? describeAppName(app.name) : 'Unknown app'
```
To double check, we are only using the list of installed applications (from Connect) for this, right? If someone, for some reason, uninstalled an app later on, it would appear as "Unknown app"?

I think this is fine, but I'll keep a mental note about it and also ask the Connect team about making sure we can detect uninstalled apps.

fabriziovigevani replied:
> To double check, we are only using the list of installed applications (from Connect) for this, right? If someone, for some reason, uninstalled an app later on, it would appear as "Unknown app"?

Yes that's correct.

> I think this is fine, but I'll keep a mental note about it and also ask the Connect team about making sure we can detect uninstalled apps.

Cool, we can add a comment and ask connect team yes :)

fabriziovigevani commented on src/components/Disputes/DisputeInfoContent.js:
```diff
@@ -296,7 +298,10 @@ const DisputedActionNA = () => {
         N/A
       </span>
-      <Help hint="">The disputed Action</Help>
+      <Help hint="">
+        This dispute is simply between the two arguments. There's no executable
+        action involved.
+      </Help>
```
Not sure if this is enough information for the user or if it's formal enough. Suggestions are welcome

sohkai replied:
Could you leave a screenshot for this case? I was looking at the deployed version to Staging, but couldn't find a dispute like this (should dispute 26 show this?).

fabriziovigevani replied:
Hmm, it should show properly
 
![Screen Shot 2020-08-07 at 3 08 51 PM](https://user-images.githubusercontent.com/22663232/89675307-15d7b500-d8c0-11ea-82a1-38702d6da340.png)

fabriziovigevani commented on src/disputables/index.js:
```diff
@@ -25,10 +25,18 @@ const DELAY_APP_ID =
 const VOTING_APP_ID =
   '0x09cdc3e6887a0002b11992e954a40326a511a1750a2f5c69d17b8b660b0d337a' // disputable-voting.aragonpm.eth
-// '0x705b5084c67966bb8e4640b28bab7a1e51e03d209d84e3a04d2a4f7415f93b34' // disputable-voting.open.aragonpm.eth
+const VOTING_OPEN_APP_ID =
+  '0x705b5084c67966bb8e4640b28bab7a1e51e03d209d84e3a04d2a4f7415f93b34' // disputable-voting.open.aragonpm.eth
```
Added support for the `open` version as well of disputable voting since that's where Facu was testing.

fabriziovigevani commented on src/hooks/useEvidences.js:
```diff
@@ -89,5 +89,16 @@ export default function useEvidences(rawEvidences) {
 
-  return [evidences, fetchingEvidences]
+  // First evidence submitted by defendant is treated as the dispute description
+  const processedEvidences = useMemo(() => {
+    if (dispute.disputable) {
+      return evidences.filter(
+        e => e.metadata !== dispute.disputable.actionContext
+      )
+    }
+
+    return evidences
+  }, [dispute.disputable, evidences])
```
From the convo offline, removed the defendant's `Justification` as argument.

sohkai commented on src/utils/dispute-utils.js:
```diff
@@ -44,6 +44,20 @@ export function transformRoundDataAttributes(round) {
+
+  return {
+    ...disputable,
+    actionContext: toUtf8String(disputable.actionContext || '0x'),
```
Right, we weren't able to decode the string from the subgraph? Was wondering if we should also use the `true` flag like we do in `parseMetadata()`?

fabriziovigevani replied:
@facuspagnuolo just updated the subgraph where it's decoded from the subgraph, so just updated the PR.

See https://github.com/aragon/court-subgraph/pull/75

sohkai commented on src/components/Disputes/DisputeInfoContent.js:
```diff
@@ -166,37 +161,151 @@ function Field({ label, loading, value, ...props }) {
-      <Link external href={disputedActionURL}>
-        {disputedActionText}
+  const ActionTextComponent = useMemo(() => {
+    // Disputes could not have an executable action
```
```suggestion
    // Disputes may not include an embedded executable action
```

sohkai commented on src/components/Disputes/DisputeInfoContent.js:
```diff
@@ -296,7 +298,10 @@ const DisputedActionNA = () => {
       </span>
-      <Help hint="">The disputed Action</Help>
+      <Help hint="">
+        This dispute is simply between the two arguments. There's no executable
+        action involved.
```
```suggestion
        This dispute does not include a binding action and is simply between the given context and arguments.
```

fabriziovigevani commented:
@sohkai there's some new changes for review.

Since now we cannot get the disputable voting `evmScript` via contract call, decided to implement a function for each disputable app which handles how it should fetch the evm script. This `extractor` function is added to each disputable app in  the `DISPUTABLE_ACTIONS` mapping. 

For disputable voting we are using `@aragon/connect-thegraph` library and performing the direct query as opposed to using the disputable voting connector. 

Also moved the disputable actions mapping to its own file to tidy things up a bit.
 
The main changes are: 

- Implementing a `scriptExtractor` function  by disputable app which is added to the `DISPUTABLE_ACTIONS` mapping.
- Upgraded  `@aragon/connect` to `latest` as i was experiencing the same issue we had when working on the `honey-pot` with the older version.
- In consequence of the previous point, had to upgrade ethers to v5 as well.

sohkai commented on src/lib/web3-utils.js:
```diff
@@ -1,17 +1,18 @@
 import { InvalidURI, InvalidNetworkType, NoConnection } from '../errors'
 import { validHttpFormat } from './uri-utils'
 
+const { id: hash256, solidityKeccak256: soliditySha3, toUtf8String } = utils
```
Oh interesting, what is `hash256` if it's not `keccak256`? (Function signatures should be using keccak256)

fabriziovigevani replied:
Ah you are right https://docs.ethers.io/v5/api/utils/hashing/#utils-id

Bad naming 😄

sohkai commented on src/disputables/queries.js:
```diff
@@ -0,0 +1,17 @@
+import gql from 'graphql-tag'
+import { GraphQLWrapper } from '@aragon/connect-thegraph'
```
I'm out of the loop on this one—is this better than the previous graphql transport we were using against the court subgraph?

fabriziovigevani replied:
We could actually just do what the library does (creating the Client and performing the query) ourselves.  
It's using the same library internally.

sohkai commented on src/disputables/scriptExtractors.js:
```diff
@@ -0,0 +1,65 @@
+export async function votingExtractor(disputableAddress, disputableActionId) {
+  // Disputable voting now saves the hash of the evmScript so we need to get it from the subgraph.
+  const networkType = sanitizeNetworkType(getNetworkType())
+  const subgraphUrl = `https://api.thegraph.com/subgraphs/name/aragon/aragon-dvoting-${networkType}${
```
Ah, could we move this to be a dynamic utility function based on the app id? We'll have a different subgraph for the precedence campaign apps.

sohkai commented on src/disputables/scriptExtractors.js:
```diff
@@ -0,0 +1,65 @@
+    SUBGRAPH_NAME ? `-${SUBGRAPH_NAME}` : ''
+  }`
+
+  const { data } = await performQuery(subgraphUrl, disputableVotingQuery, {
```
I wonder if it would be a better API to hide `performQuery()` and only expose `performDisputableVotingQuery(id, voteId)`. Otherwise it might be a bit hard to know which params to use?

fabriziovigevani replied:
Yes, good one!

sohkai reviewed (approved):
I think we should also be able to bump the bundle-checker down now, since we should be deduping ethers versions!

I otherwise have a couple of comments on things I know we'll want to generalize further down the line. Otherwise, looking good to me!

sohkai commented on src/disputables/index.js:
```diff
@@ -229,27 +234,38 @@ function buildDisputedActionRadspec(transactions) {
+
+  let url
+  if (arbitrable) {
+    url = [arbitrable.link, arbitrable.entityPath, actionId].join('/')
```
Not a big deal, but maybe this is something we would hide inside the API since each arbitrable could have its own URL scheme?

Perhaps something like `isArbitrableKnown(arbitrableAddress, networkType) && buildArbitrableUrl(arbitrableAddress, actionId, networkType)` would be better? And then the mapping would be from `address => function`.

sohkai commented on src/disputables/queries.js:
```diff
@@ -1,9 +1,21 @@
 import gql from 'graphql-tag'
-import { GraphQLWrapper } from '@aragon/connect-thegraph'
+import { Client } from 'urql/'
```
```suggestion
import { Client } from 'urql'
```

Not sure, but did we need the `/`?

fabriziovigevani replied:
Oops, no we don't
